sql-plugin/src/main/scala/com/nvidia/spark/rapids/FoldLocalAggregate.scala

```diff
@@ -0,0 +1,128 @@
+        finalAgg match {
+          case hash: HashAggregateExec =>
+            hash.copy(
+              groupingExpressions = partAgg.groupingExpressions,
+              aggregateExpressions = aggExpressions,
+              aggregateAttributes = aggAttributes,
+              child = partAgg.child)
+          case sort: SortAggregateExec =>
+            sort.copy(
+              groupingExpressions = partAgg.groupingExpressions,
+              aggregateExpressions = aggExpressions,
+              aggregateAttributes = aggAttributes,
+              child = partAgg.child)
+          case obj: ObjectHashAggregateExec =>
+            obj.copy(
+              groupingExpressions = partAgg.groupingExpressions,
+              aggregateExpressions = aggExpressions,
+              aggregateAttributes = aggAttributes,
+              child = partAgg.child)
+          case _ =>
+            throw new IllegalStateException(s"Unexpected AggregateExec: $finalAgg")
+        }
+    }
+  }
+}
+
+/**
+ * Capture the pattern of LocalAggregate, which consists of two-stage AggregateExecs being
+ * connected with each other directly (without the ShuffleExchange in between):
+ * +- FinalAggregateExec
+ *    +- PartialAggregateExec
+ *
+ * The LocalAggregate can be emerged regardless HashAggregateExec, SortAggregateExec or
+ * ObjectHashAggregateExec.
+ */
+object LocalAggregatePattern extends Logging {
+  def unapply(plan: SparkPlan): Option[(BaseAggregateExec, BaseAggregateExec)] = {
+    plan match {
+      case hashAgg: HashAggregateExec
+        if hashAgg.child.isInstanceOf[HashAggregateExec] &&
+          isLocalExchange(hashAgg, hashAgg.child.asInstanceOf[HashAggregateExec]) =>
+        Some(hashAgg -> hashAgg.child.asInstanceOf[HashAggregateExec])
+      case sort: SortAggregateExec
+        if sort.child.isInstanceOf[SortAggregateExec] &&
+          isLocalExchange(sort, sort.child.asInstanceOf[SortAggregateExec]) =>
+        Some(sort -> sort.child.asInstanceOf[SortAggregateExec])
+      case obj: ObjectHashAggregateExec
+        if obj.child.isInstanceOf[ObjectHashAggregateExec] &&
+          isLocalExchange(obj, obj.child.asInstanceOf[ObjectHashAggregateExec])  =>
+        Some(obj -> obj.child.asInstanceOf[ObjectHashAggregateExec])
+      case _ =>
+        None
+    }
+  }
+
+  // Check if the two connected AggregateExecs are two stages of the same logical Aggregate. If
+  // so, group expressions and aggregate expressions should be identical(except AggregateMode)
+  private def isLocalExchange(merge: BaseAggregateExec, partial: BaseAggregateExec): Boolean = {
+    val groupExpressions = merge.groupingExpressions
+    val aggExpressions = merge.aggregateExpressions
+    val childGroupExpressions = partial.groupingExpressions
+    val childAggExpressions = partial.aggregateExpressions
+    // Fast path
+    if (groupExpressions.length != childGroupExpressions.length ||
+      aggExpressions.length != childAggExpressions.length) {
+      return false
```

Comment:
Could we write this to avoid `return` in scala?  Same comment for the other two returns.

Reply:
Done.